Add reverse binomial with tests.

hernanmd opened the pull request:
See https://math.stackexchange.com/a/103385/205 for details

hemalvarambhia commented on src/Math-Numerical/Integer.extension.st:
```diff
@@ -1,5 +1,54 @@
+	See https://math.stackexchange.com/a/103385/205 for details. "
+
+	| k |
+	[ self isInteger and: [ self > 1 ] ] assert.
```
the method is a little long. Is there a way to break it up?

hemalvarambhia commented on src/Math-Tests-Numerical/IntegerTest.extension.st:
```diff
@@ -0,0 +1,42 @@
+	self assert: 3003 findNK equals: #(3003 1 3003 3002 78 2 78 76 15 5 15 10 14 6 14 8).
+	self assert: 8966473191018617158916954970192684 findNK equals: #(8966473191018617158916954970192684 1 8966473191018617158916954970192684 8966473191018617158916954970192683 123 45 123 78).
+	self assert: 116682544286207712305570174244760883486876241791210077037133735047856714594324355733933738740204795317223528882568337264611289789138133946725471443924278812277695432803500115699090641248357468388106131543801393801287657125117557432072695477147403395443757359171876874010770591355653882772562301453205472707597435095925666815012707478996454360460481159339802667334477440 findNK equals:  ...
+	self should: [ 1 findNK ] raise: Error.
```
Move these to test methods as proposed by @olekscode perhaps?

hemalvarambhia commented on src/Math-Numerical/Integer.extension.st:
```diff
@@ -1,5 +1,54 @@
 Extension { #name : #Integer }
 
+{ #category : #'*Math-Numerical' }
+Integer >> findNK [
```
Maybe `reverseBinomial:` given your comment  in the method block?

hemalvarambhia commented on src/Math-Numerical/Integer.extension.st:
```diff
@@ -1,5 +1,54 @@
+	See https://math.stackexchange.com/a/103385/205 for details. "
+
+	| k |
+	[ self isInteger and: [ self > 1 ] ] assert.
```
Given you've defined this in the Integer class, do you need to have the `isInteger` check?

hernanmd replied:
You're right! I removed the check. Thanks!

hemalvarambhia commented on src/Math-Numerical/Integer.extension.st:
```diff
@@ -1,5 +1,54 @@
+]
+
+{ #category : #'*Math-Numerical' }
+Integer >> firstOver: c [
```
Maybe add a link to the underlying Mathematics? This is a part of the domain I'm not clear on.

hemalvarambhia commented:
As a first observation this seems to be a direct translation of the python code snippet in the stack exchange link.

hemalvarambhia commented:
@hernanmd as I mentioned above, the code appears to be python translated, almost literally to Smalltalk. Is this the case? 

I think `findNK` could be made more intention revealing. Could the code be refactored a little as I can't seem to follow it too easily (I'm having an off day!).

hernanmd commented:
> @hernanmd as I mentioned above, the code appears to be python translated, almost literally, to Smalltalk. Is this the case?
> 
> I think `findNK` could be made more intention revealing. Could the code be refactored a little as I can't seem to follow it too easily (I'm having an off day!).

Yes, it is a translation from Python code that's why I included the link :)

I don't see how it could be refactored because Integer can easily explode.

nicolas-cellier-aka-nice commented on src/Math-Tests-Numerical/IntegerTest.extension.st:
```diff
@@ -0,0 +1,42 @@
+	self assert: (4 firstOver: 4) equals: 2.
+	self assert: (4 firstOver: 5) equals: 2.
+	
+	self assert: (4 firstOver: SmallInteger maxVal) equals: 32768.
```
SmallInteger maxVal will vary in 32bits vs 64bits image flavours...
Please use an explicit value

hernanmd replied:
Done, thanks!

hemalvarambhia commented:
Chasing up on this PR, @hernanmd . Let us know when you've addressed review comments and then we can merge

hemalvarambhia commented:
@hernanmd chasing up again :)

nicolas-cellier-aka-nice commented on src/Math-Numerical/Integer.extension.st:
```diff
@@ -1,5 +1,54 @@
+]
+
+{ #category : #'*Math-Numerical' }
+Integer >> firstOver: c [
```
firstOver: is not really intention revealing...

If I understand, that's solving `n^self>=c`, that is `n=(c^(1/self)) ceiling`, or in other words, that's `(c nthRoot: self) ceiling`.
Both Cuis and Squeak have `nthRoot:`. If Pharo lacks such method, then it shall be added to Pharo.

hernanmd replied:
I quickly checked but `2 nthRoot: 3`in Squeak has different output than `2 firstOver: 3`.

SergeStinckwich commented:
Move to 1.0.4 milestone

SergeStinckwich commented:
Can we merge? @hemalvarambhia

hemalvarambhia commented:
@SergeStinckwich I think the code quality may be improved. Plus, there has been no addressing of comments or any other activity. 

I would close it rather than merge. That way, it can reopened when the author has energy and capacity. 

What do other reviewers think?

hernanmd commented:
> @SergeStinckwich I think the code quality may be improved. Plus, there has been no addressing of comments or any other activity.

Can you detail which aspect of the code could be improved?

IMHO it seems too much revision for a single method addition, I would love to know why or simply discard this addition definitely.

SergeStinckwich commented:
The main concern of @hemalvarambhia is about names. Could you use better names for ``findNK`` and ``firstOver:`` ?

SergeStinckwich commented:
I propose to replace ``findNK`` by ``findAllNKGivenBinomialCoeff``or ``inverseBinomialCoefficient``.
You can remove ``firstOver:``and replace directly in the code: ``nmin := k firstOver: (k factorial * self).``by ``nmin :=  (k factorial * self) nthRoot: k.``.

hemalvarambhia commented:
> I propose to replace ``findNK`` by ``findAllNKGivenBinomialCoeff``or ``inverseBinomialCoefficient``.
> You can remove ``firstOver:`` (and tests) and replace directly in the code: ``nmin := k firstOver: (k factorial * self).``by ``nmin :=  (k factorial * self) nthRoot: k.``

`inverseBinomialCoeffient` feels more concise IMHO. Could you use Kent Beck's 'Refactor to composed method' to improve the maintainability. This is the only concern that I have.

hemalvarambhia commented:
@hernanmd This is only my opinion and it is important that others weigh in. If more disagree with my view, ignore the following:

my concern is the structure of `findNK`. It feels very nested, with `while` and nested `ifTrue` messages so I sense there is a risk to maintainability and understandability, a key trade off for me. The method is also long. One option to improve things might be to appeal to 'refactor to composed method' (cf Kent Beck's 'Smalltalk Best Practice Pattern')

SergeStinckwich commented:
Thank you Hernan.
